plugins/modules/ipa_pwpolicy.py

```diff
@@ -34,36 +34,56 @@
         type: str
     maxpwdlife:
         description: Maximum password lifetime (in days).
-        type: str
+        type: int
     minpwdlife:
         description: Minimum password lifetime (in hours).
-        type: str
+        type: int
     historylength:
         description:
         - Number of previous passwords that are remembered.
         - Users cannot reuse remembered passwords.
-        type: str
+        type: int
     minclasses:
         description: Minimum number of character classes.
-        type: str
+        type: int
     minlength:
         description: Minimum password length.
-        type: str
+        type: int
     priority:
         description:
         - Priority of the policy.
         - High number means lower priority.
         - Required when C(cn) is not the global policy.
-        type: str
+        type: int
     maxfailcount:
         description: Maximum number of consecutive failures before lockout.
-        type: str
+        type: int
     failinterval:
         description: Period (in seconds) after which the number of failed login attempts is reset.
-        type: str
+        type: int
     lockouttime:
         description: Period (in seconds) for which users are locked out.
-        type: str
+        type: int
```

Comment:
Changing types of existing options could be potential breaking changes. Was it possible to pass any other string for these values that were not string representations of integers, or did the API did not accept them before?

Also how does this affect idempotency: does idempotency still work? Or does that one needs to compare the values as strings?

Also, finally, what happens if the strings have leading zeros, or prefixes such as `0x`, `0o`, etc? Did the API accept any of these values? If yes, does changing the type change the behavior?

Reply:
I have tested these changes on my local environment and it's seems okay.

This changes are backward compatible and if the values are in string the module will still work with idempotency.

The module is idempotent and will show no changes after multiple apply and compatible on the change between integers or strings.

In case of using strings the string must be convertible to integers and according the Freeipa documentation it's only accept integers not `0x`, `0o`.
```{"changed": false, "msg": "argument 'failinterval' is of type <class 'str'> and we were unable to convert to int: <class 'str'> cannot be converted to an int"}```

If someone use leading zero integers the module will fail in idempotency. I can add another commit to fix this but I think it's extra and kind of user behavior.

Reply:
> In case of using strings the string must be convertible to integers and according the Freeipa documentation it's only accept integers not `0x`, `0o`. `{"changed": false, "msg": "argument 'failinterval' is of type <class 'str'> and we were unable to convert to int: <class 'str'> cannot be converted to an int"}`

This message looks like it comes from ansible-core when a string is provided to a `type: int` option. Are you sure that this error came from FreeIPA and not from ansible-core after you changed `type: str` to `type: int`?

Reply:
Yes, It's ansible error, I have tested FreeIpa and they accept hex values too like `0x` or `0x3` and if someone use hex values in their ansible variables this change will break their play.

Any suggestion? keep it in integers or revert these values to strings!

Reply:
In my opinion and according the FreeIpa documention we should keep it integers, PTAL at this [link](https://freeipa.readthedocs.io/en/latest/api/pwpolicy_add.html)

Reply:
While I agree that would be nicer, we still do not accept breaking changes. If the module accepted values such as `0x3` before, we still need to do that after this change.

Reply:
@felixfontein PTAL! I think we can keep it in integers because when using string variables like `0x3` the module will not idempotent and will throw error in second run.

In the `get_diff` function at `ipa.py` the `module_data` is `0x3` but the `ipa_data`  will be `3` and the module try to modify pw_policy and the modification will fail with below error:
```
fatal: [localhost]: FAILED! => {"changed": false, "msg": "response pwpolicy_mod: no modifications to be performed"}
```

get_diff function
```
    def get_diff(self, ipa_data, module_data):
        result = []
        for key in module_data.keys():
            mod_value = module_data.get(key, None)
            if isinstance(mod_value, list):
                default = []
            else:
                default = None
            ipa_value = ipa_data.get(key, default)
            if isinstance(ipa_value, list) and not isinstance(mod_value, list):
                mod_value = [mod_value]
            if isinstance(ipa_value, list) and isinstance(mod_value, list):
                mod_value = sorted(mod_value)
                ipa_value = sorted(ipa_value)
            if mod_value != ipa_value:
                result.append(key)
        return result
```

For better understanding here is an output of the `ipa_data` and `module_data` , take a look at `passwordgracelimit` as an example:
```
{'krbmaxpwdlife': '90', 'krbpwdhistorylength': '1', 'krbpwdmindiffchars': '2', 'krbpwdminlength': '12', 'krbpwdmaxfailure': '5', 'krbpwdlockoutduration': '300', 'passwordgracelimit': '0x3', 'ipapwddictcheck': False, 'ipapwdusercheck': True}

{'objectclass': ['top', 'nsContainer', 'krbPwdPolicy', 'ipaPwdPolicy'], 'ipapwdusercheck': [True], 'krbpwdmindiffchars': ['2'], 'krbpwdminlength': ['12'], 'krbpwdfailurecountinterval': ['60'], 'krbpwdmaxfailure': ['5'], 'krbpwdlockoutduration': ['300'], 'krbminpwdlife': ['1'], 'passwordgracelimit': ['3'], 'ipapwddictcheck': [False], 'krbmaxpwdlife': ['90'], 'krbpwdhistorylength': ['1'], 'cn': ['global_policy'], 'ipapwdmaxsequence': ['3'], 'dn': 'cn=global_policy,cn=TEST,cn=kerberos,dc=test,dc=test'}
```

Reply:
> @felixfontein PTAL! I think we can keep it in integers because when using string variables like `0x3` the module will not idempotent and will throw error in second run.

It's still a breaking change and won't get merged.

Reply:
I understand your concern about the breaking change. AFIK it's not breaking because:
1. module will not work properly with `0x3` as an string value because this will fail in multiple run as I mentioned in previous reply.
2. FreeIPA documentation clearly define these values as integers and not string
3. If someone currently use strings like '1' or '2' the change will not break their play

BTW if you are persist about the breaking change I can revert these values to strings.

Reply:
It does not fail/misbehave on the first call, so it's still breaking behavior for some use-cases.

I'm all for changing this, **but** it has to be done in a backwards-compatible way. That means that these values still have to be accepted, but a deprecation message has to be printed (`module.deprecate()`), so that users are informed and can adjust their playbooks.

Reply:
I concur with @felixfontein comments and approach here.  While I understand the desire to make it 'right', now that it is in the wild and has been for years against a lot of different IPA versions and probably weird use cases, there is a balance between making it correct and keeping it consistent.  This is one of those times where being consistent is more important than being correct.  There appears to be little reward for risking the change.  Your arguments for it being fine/safe to change also cut both ways: the current methods work fine for your use case as well, near as I can tell.  With an uncountable user base (near as I can tell) for this module and no visibility into their implementations, I see no reason to risk a breaking change here.

Reply:
Thanks @justchris1 , @felixfontein, I understand the limitations.

@felixfontein PTALA.